File: problem_set_3/longest_increasing_subsequence.go
package main

import "fmt"

func lengthOfLIS(nums []int) int {
	end := len(nums)
	database := make([]int, end)

    for i := range database {
        database[i] = 1
    }

	for idx := 1; idx < end; idx++ {
		for counter := 0; counter < idx; counter++ {
			if nums[idx] > nums[counter] {
				database[idx] = max(database[idx], database[counter]+1)
			}
		}
	}

	return max(database...)
}

func max(nums ...int) int {
	result := nums[0]
	for _, num := range nums {
		if num > result {
			result = num
		}
	}
	return result
}

func main() {
	fmt.Println(lengthOfLIS([]int{10, 9, 2, 5, 3, 7, 101, 18}))
	fmt.Println(lengthOfLIS([]int{0, 1, 0, 3, 2, 3}))
	fmt.Println(lengthOfLIS([]int{1, 1, 1, 1, 1, 1, 1}))
}

File: problem_set_3/longest_increasing_subsequence.py
from typing import List

def lengthOfLIS(nums: List[int]) -> int:
    
    end = len(nums)
    database = [1] * end

    for idx in range(1, end):
        counter = 0
        while counter != idx:
            if nums[idx] > nums[counter]:
                database[idx] = max(database[idx], database[counter] + 1)
            counter+=1

    return max(database)


if __name__ == '__main__':
    print(lengthOfLIS([10, 9, 2, 5, 3, 7, 101, 18]))
    print(lengthOfLIS([0, 1, 0, 3, 2, 3]))
    print(lengthOfLIS([1, 1, 1, 1, 1, 1, 1]))